### PR TITLE
refactor PagedMergeSort

### PR DESCRIPTION
**Pull request by LilithHafner**
I feel like these suggestions are a bit to comprehensive to put in github line-by-line-comments so I made a PR to your topic branch, @LSchwerdt.

I suggest these substantial changes:
- Remove the 1,2,3 from the beginning of the `pageLocation` array and move the final partial page directly into position, removing it from `pageLocation` as well, shortening the array by a total of 4 and helping simplify code.
- Shift responsibility to copy data back from the buffer from `copy_pages!` to the caller which allows that function to be substantially simpler
- Refactor how the merge is handled once one of `A` or `B` is close to empty theoretically resulting in minor (negligible) performance improvement and also simplifying that code and setting up for "calculate location of the 3 free pages" to be much simpler
- Pretty much eliminate the "calculate location of the 3 free pages" section altogether
- Switch from a while loop to a slightly more concise for loop when doing the final unresolved cycle detection pass. Also skip the first 4 blocks because the first three are already in place and any nontrivial cycle will appear in at least 2 positions so skipping one is okay.
- remove `copy_page!(v, source, offset, offset2, pagesize)` and replace its usages with `copyto!` which is equivalent for `Vectors` but may have more efficient implementations for other `AbstractVectors`

And these more superficial changes:
- inline assignments of the form
```julia
x = 1
f(x)
```
- inline internal functions `next_page_A` and `next_page_B` that are only called once
- replace unused values in unpacking with `_`

### CODE REVIEW

File: src/SortingAlgorithms.jl [hunk 1]
```diff
@@ -852,119 +832,72 @@ function paged_merge!(v::AbstractVector{T}, lo::Integer, m::Integer, hi::Integer
...
-            # nFreePagesA == 3 is impossible:
-            # B contains at least 3pagesize elements -> at least one free page will exist in B at some point
-            # next_page!() never uses nextB if there is more than one page free in A
-            if nFreePagesA == 1
-                freePages = (pages.nextA, pages.nextB, pages.nextB + 1)
-            else # nFreePagesA == 2
-                freePages = (pages.nextA, pages.nextA + 1, pages.nextB)
+            use_a = a <= m
+            # copy the incomplete page
+            amt = offset + pagesize - k + 1
```
Comment (LSchwerdt) [inline src/SortingAlgorithms.jl]: I dont like `amt` as variable name. The Julia style discourages abbreviated words as variable names, and 
`amount` is also not very clear. I prefer `partial_page_size`, but maybe you have a better idea.

File: src/SortingAlgorithms.jl [hunk 2]
```diff
@@ -782,42 +780,24 @@ struct Pages
...
 end
 
-next_page_A(pages::Pages) = Pages(pages.nextA, pages.currentNumber + 1, pages.nextA + 1, pages.nextB)
-next_page_B(pages::Pages) = Pages(pages.nextB, pages.currentNumber + 1, pages.nextA, pages.nextB + 1)
-
 Base.@propagate_inbounds function next_page!(pageLocations, pages, pagesize, lo, a)
-    if a > pages.nextA * pagesize + lo
-        pages = next_page_A(pages)
+    pages = if a > pages.nextA * pagesize + lo
+        Pages(pages.nextA, pages.currentNumber + 1, pages.nextA + 1, pages.nextB)
```
Comment (LSchwerdt) [inline src/SortingAlgorithms.jl]: The functions `next_page_A/B` were kept on purpose to help with understanding `next_page!`. If you insist on removing the functions, I would add comments like this:
```julia
pages = if a > pages.nextA * pagesize + lo
    # use next page from A
    Pages(pages.nextA, pages.currentNumber + 1, pages.nextA + 1, pages.nextB) 
else
    # use next page from B
    Pages(pages.nextB, pages.currentNumber + 1, pages.nextA, pages.nextB + 1)
end
```
This eliminates the need to understand how the `Pages` Type is laid out to understand the function `next_page!`. But I prefer keeping the functions.

File: src/SortingAlgorithms.jl [hunk 3]
```diff
@@ -852,119 +832,72 @@ function paged_merge!(v::AbstractVector{T}, lo::Integer, m::Integer, hi::Integer
...
+                offset = page_offset(pages.current)
+                copyto!(v, offset+1, v, use_a ? a : b, pagesize)
+                use_a && (a += pagesize)
+                use_a || (b += pagesize)
             end
+            # copy the final partial page only if sourcing from A.
+            # If sourcing from B, it is already in place.
+            use_a && copyto!(v, hi - m + a, v, a, m - a + 1)
         end
+        # return vcat(v, 40, buf)
```
Comment (LSchwerdt) [inline src/SortingAlgorithms.jl]:
```suggestion

```

**Comment by LSchwerdt** [comment]
Amazing work! Sorry for taking so long for my comments. But I only have some minor remarks.

File: src/SortingAlgorithms.jl (hunk 1 above)
Comment (LilithHafner) [reply src/SortingAlgorithms.jl]: Absolutely, thanks. I've changed it to `partial_page_size`.

File: src/SortingAlgorithms.jl (hunk 3 above)
Comment (LilithHafner) [reply src/SortingAlgorithms.jl]: Thanks!

File: src/SortingAlgorithms.jl (hunk 2 above)
Comment (LilithHafner) [reply src/SortingAlgorithms.jl]: I have mixed feelings on this, but I think you may be right. I'll put back the functions.

**Pull request merged**